### ISSUE TITLE
C-array's shape deduced wrongly for certain types

### ISSUE THREAD

**Issue by sbacchio** [wlav/cppyy#5]
Hi,

the shape of c-style arrays is deduced wrongly for various types.
This happens for instance for `char`, `enum`, `void*` and probably others.

Here a way to reproduce the issue:

```python
import cppyy

cppyy.cppdef("""
enum Enum{One, Two, Three};

template<typename T>
struct Foo {
  T a[5];
  T aa[5][4];
  T aaa[5][4][3];
};
""")

for dtype in ["int", "double", "long", "bool", "char", "void*", "Enum"]:
    foo = cppyy.gbl.Foo[dtype]()
    print(dtype, foo.a.shape, foo.aa.shape, foo.aaa.shape)
```

And the output with `cppyy==2.1.0` is

```
int (5,) (5, 4) (5, 4, 3)
double (5,) (5, 4) (5, 4, 3)
long (5,) (5, 4) (5, 4, 3)
bool (5,) (5, 4) (5, 4, 3)
char (1,) (268435455,) (268435455,)
void* (5,) (268435455,) (268435455,)
Enum (5,) (268435455,) (268435455,)
```

Clearly one would expect for any datatype the shapes `(5,) (5, 4) (5, 4, 3)`. As you can see, though, for `enum` and `void*` it fails to deduce the shape of array of array, while for `char` already the shape of a single array is wrong being (1,) and not (5,).

Thanks a lot in advance!

**Comment by wlav** [comment]
Looks like the wrong info comes from the backend, with in principle no difference in treatment for the types once they hit the converters (there has been a significant overhaul of array argument handling for 2.1.0).

`268435455` is `2**31/8-1`, i.e. signed max ('unknown') size for 8 byte elements (`sizeof(void*) == 8`).

Will do some digging, but am a bit overloaded atm. (just back online after jury duty).

**Comment by wlav** [comment]
That was hairy, but is fixed in repo now. The only caveat is that `char[5]` is still a Python `str`, but it will be one of length 5 (as opposed to a LowLevelView of `char` of length 5, or a `str` up to '\0' as was the alternative). Improvements in determining how `char`/`str` should be handled are waiting on a fix to this bug: https://bitbucket.org/wlav/cppyy/issues/344/failed-to-handle-char-array-with-arbitrary

**Comment by sbacchio** [comment]
Thanks for looking at it!

**Comment by wlav** [comment]
Released with cppyy 2.2.0 and its dependencies.
